sklearn/metrics/_classification.py

```diff
@@ -2463,8 +2475,12 @@ def balanced_accuracy_score(y_true, y_pred, *, sample_weight=None, adjusted=Fals
     with np.errstate(divide="ignore", invalid="ignore"):
         per_class = np.diag(C) / C.sum(axis=1)
     if np.any(np.isnan(per_class)):
-        warnings.warn("y_pred contains classes not in y_true")
-        per_class = per_class[~np.isnan(per_class)]
+        if zero_division == "warn":
+            warnings.warn("y_pred contains classes not in y_true")
```

Comment:
We need to raise a `UndefinedMetricWarning` and not a `UserWarning` for consistency.

Reply:
Here, we should raise the following message:

```
balanced_accuracy ill-defined and being set to 0.0.
Use `zero_division` parameter to control this behaviour.
```

Reply:
We can also gain a bit of space

```python
nan_replacement_value = zero_division if zero_division != "warn" else 0.0
per_class = np.nan_to_num(per_class, nan=nan_replacement_value)
if zero_division == "warn":
    warnings.warn(...)
```